btcec/signature.go

```diff
@@ -396,3 +401,134 @@ func RecoverCompact(curve *KoblitzCurve, signature,
 
 	return key, ((signature[0] - 27) & 4) == 4, nil
 }
+
+var one = big.NewInt(1)
+
+// SignRFC6979 generates a deterministic ECDSA signature according to RFC 6979 and BIP 62.
+func SignRFC6979(privateKey *PrivateKey, hash []byte) (*Signature, error) {
+
+	privkey := privateKey.ToECDSA()
+	N := order
+	k := NonceRFC6979(privkey.D.Bytes(), hash)
+	inv := new(big.Int).ModInverse(k, N)
+	r, _ := privkey.Curve.ScalarBaseMult(k.Bytes())
+	r.Mod(r, N)
+
+	if r.Sign() == 0 {
+		return nil, errors.New("calculated R is zero")
+	}
+
+	e := hashToInt(hash, privkey.Curve)
+	s := new(big.Int).Mul(privkey.D, r)
+	s.Add(s, e)
+	s.Mul(s, inv)
+	s.Mod(s, N)
+
+	if s.Cmp(halforder) == 1 {
+		s = new(big.Int).Sub(order, s)
+	}
+
+	return &Signature{R: r, S: s}, nil
+}
+
+// NonceRFC6979 generates a ECDSA nonce (`k`) deterministically according to RFC 6979.
+// It takes a 32-byte hash as an input and returns 32-byte nonce to be used in ECDSA algorithm.
+func NonceRFC6979(privkey []byte, hash []byte) *big.Int {
```

Comment:
Is there any reason this is exported?  I see it's called in the tests, but if that's the only reason for it, it should be exported via the `internal_test.go` so it's only exported during the tests instead of all the time.